internal/intern/intern_test.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// package intern provides an interning table abstraction to optimize symbol
+// resolution.
+package intern_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/bufbuild/protocompile/internal/intern"
+)
+
+func TestIntern(t *testing.T) {
+	t.Parallel()
+
+	data := []string{
+		"",
+		"a",
+		"abc",
+		"?",
+		"xy.z",
+		"a_b_c",
+		".....",
+		"foo.",
+		"foo.a",
+		"very long",
+		" ",
+		"verylong",
+	}
+
+	var table intern.Table
+	for i := 0; i < 3; i++ {
+		for _, s := range data {
+			s := s
+
+			t.Run(fmt.Sprintf("%s/%d", s, i), func(t *testing.T) {
+				t.Parallel()
+
+				id := table.Intern(s)
+				assert.Equal(t, s, table.Value(id), "id: %v", id)
```

Comment:
Maybe an additional assertion that strings whose length is <= 5 result in a negative ID?